sdk/src/types/block/output/feature/metadata.rs

```diff
@@ -27,77 +59,108 @@ impl MetadataFeature {
+                return Err(Error::NonGraphicAsciiMetadataKey(k.to_vec()));
             }
+            if res
+                .insert(
+                    k.into_boxed_slice()
+                        .try_into()
+                        .map_err(Error::InvalidMetadataFeatureKeyLength)?,
+                    v.into_boxed_slice()
+                        .try_into()
+                        .map_err(Error::InvalidMetadataFeatureValueLength)?,
+                )
+                .is_some()
+            {
+                return Err(Error::InvalidMetadataFeature("Duplicated metadata key".to_string()));
+            };
         }
-    };
-}
-impl_from_vec!(&str);
-impl_from_vec!(String);
-impl_from_vec!(&[u8]);
-
-impl<const N: usize> TryFrom<[u8; N]> for MetadataFeature {
-    type Error = Error;
-
-    fn try_from(value: [u8; N]) -> Result<Self, Self::Error> {
-        value.to_vec().try_into()
-    }
-}
-
-impl TryFrom<Vec<u8>> for MetadataFeature {
-    type Error = Error;
-
-    fn try_from(data: Vec<u8>) -> Result<Self, Error> {
-        data.into_boxed_slice().try_into()
+        Ok(Self(res.try_into().map_err(Error::InvalidMetadataFeatureLength)?))
     }
 }
 
-impl TryFrom<Box<[u8]>> for MetadataFeature {
+impl TryFrom<BTreeMap<Vec<u8>, Vec<u8>>> for MetadataFeature {
     type Error = Error;
 
-    fn try_from(data: Box<[u8]>) -> Result<Self, Error> {
-        data.try_into().map(Self).map_err(Error::InvalidMetadataFeatureLength)
-    }
-}
-
-impl FromStr for MetadataFeature {
-    type Err = Error;
-
-    fn from_str(s: &str) -> Result<Self, Self::Err> {
-        Self::new(prefix_hex::decode::<Vec<u8>>(s).map_err(Error::Hex)?)
+    fn try_from(data: BTreeMap<Vec<u8>, Vec<u8>>) -> Result<Self, Error> {
+        let mut res = BTreeMap::<
+            BoxedSlicePrefix<u8, MetadataFeatureKeyLength>,
+            BoxedSlicePrefix<u8, MetadataFeatureValueLength>,
+        >::new();
+        for (k, v) in data {
+            if !k.iter().all(|b| b.is_ascii_graphic()) {
+                return Err(Error::NonGraphicAsciiMetadataKey(k.to_vec()));
+            }
+            res.insert(
+                k.into_boxed_slice()
+                    .try_into()
+                    .map_err(Error::InvalidMetadataFeatureKeyLength)?,
+                v.into_boxed_slice()
+                    .try_into()
+                    .map_err(Error::InvalidMetadataFeatureValueLength)?,
+            );
+        }
+        Ok(Self(res.try_into().map_err(Error::InvalidMetadataFeatureLength)?))
     }
 }
 
 impl core::fmt::Display for MetadataFeature {
     fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
-        write!(f, "{}", prefix_hex::encode(self.data()))
+        write!(
+            f,
+            "{:?}",
+            self.0
+                .keys()
+                // Safe to unwrap, keys must be ascii
+                .map(|k| alloc::str::from_utf8(k).unwrap())
+                .collect::<Vec<&str>>()
```

Comment:
Only showing keys ?

Reply:
They're at least human readable, values would be hex encoded, not sure if we want them or not